src/api/utils.rs

```diff
@@ -19,4 +25,110 @@ pub mod utils {
+    #[napi]
+    pub async fn show_gamepad_text_input(
+        input_mode: GamepadTextInputMode,
+        input_line_mode: GamepadTextInputLineMode,
+        description: String,
+        max_characters: u32,
+        existing_text: Option<String>,
+    ) -> Option<String> {
+        let client = crate::client::get_client();
+
+        let (tx, rx) = oneshot::channel();
+        let mut tx = Some(tx);
+
+        let opened = client.utils().show_gamepad_text_input(
+            match input_mode {
+                GamepadTextInputMode::Normal => kGamepadTextInputMode::Normal,
+                GamepadTextInputMode::Password => kGamepadTextInputMode::Password,
+            },
+            match input_line_mode {
+                GamepadTextInputLineMode::SingleLine => kGamepadTextInputLineMode::SingleLine,
+                GamepadTextInputLineMode::MultipleLines => kGamepadTextInputLineMode::MultipleLines,
+            },
+            &description,
+            max_characters,
+            existing_text.as_deref(),
+            move |dismissed_data| {
+                if let Some(tx) = tx.take() {
+                    let text = client
+                        .utils()
+                        .get_entered_gamepad_text_input(&dismissed_data);
+                    tx.send(text).unwrap();
+                }
+            },
+        );
+
+        if opened {
+            rx.await.unwrap()
+        } else {
+            None
+        }
+    }
+
+    #[napi]
+    pub enum FloatingGamepadTextInputMode {
+        SingleLine,
+        MultipleLines,
+        Email,
+        Numeric,
+    }
+
+    /// @returns true if the floating keyboard was shown, otherwise, false
+    #[napi]
+    pub async fn show_floating_gamepad_text_input(
+        keyboard_mode: FloatingGamepadTextInputMode,
+        x: i32,
+        y: i32,
+        width: i32,
+        height: i32,
+    ) -> bool {
+        let client = crate::client::get_client();
+
+        let (tx, rx) = oneshot::channel();
+        let mut tx = Some(tx);
+
+        let opened = client.utils().show_floating_gamepad_text_input(
+            match keyboard_mode {
+                FloatingGamepadTextInputMode::SingleLine => {
+                    kFloatingGamepadTextInputMode::SingleLine
+                }
+                FloatingGamepadTextInputMode::MultipleLines => {
+                    kFloatingGamepadTextInputMode::MultipleLines
+                }
+                FloatingGamepadTextInputMode::Email => kFloatingGamepadTextInputMode::Email,
+                FloatingGamepadTextInputMode::Numeric => kFloatingGamepadTextInputMode::Numeric,
+            },
+            x,
+            y,
+            width,
+            height,
+            move || {
+                if let Some(tx) = tx.take() {
+                    tx.send(true).unwrap();
+                }
+            },
+        );
+
+        if opened {
+            rx.await.unwrap()
+        } else {
+            false
+        }
```

Comment:
If I'm understanding the [docs](https://partner.steamgames.com/doc/api/ISteamUtils#ShowFloatingGamepadTextInput) correctly, this function will immediately return `true`/`false` if the keyboard was shown or not, and invoke the dismissed callback once the keyboard is dismissed by user. The API might need some adjustments in that case as this promise based API doesn't behave the same way. An alternative solution would be to return `false | Promise<void>` to denote that either it wasn't opened (and there is no need to listen for the dismissed event), or a tuple/object like `[opened: boolean, onDismissed: Promise<void>]` for a cleaner state separation.

Let me know if I'm missing something.

Reply:
I can't see a real use case where you will need both, I think the promise way maes more sense on JS and can benefit the consumer for a more fluid code flow. What do you think?